### PR TITLE
Replace Ledger tests with automated tests and run them on Travis

### LINKED ISSUE #277: Integrate a Ledger simulator into CI
May be the ticket: https://github.com/LedgerHQ/speculos

### PR DESCRIPTION
**Pull request by achow101**
This PR adds the necessary build and test scripts to make use of Ledger's [Speculos](https://github.com/LedgerHQ/speculos) emulator. Speculos emulates Ledger Nano S's and will allow us to test that the commands we send to Ledger devices work.

Following the same pattern as our other device tests, Speculos will be cloned to `test/work`, built from source, and then run by the `test_ledger.py` script. The functionality to use a physical device in this test has been removed and the command line args to `test_ledger.py` and `run_tests.py` have been updated to reflect this. They follow the same pattern as other device tests.

As the emulator does not automatically accept and click through user interface prompts, we needed some way to do the user interaction. This required two pieces: 1) a way to read the emulated display, and 2) another thread to send input. To get the display output, I've added a patch that will output the text displayed to a unix socket. This way we can read the display and know which keypresses to send. Then, another thread is started which listens on that unix socket and sends the appropriate keypress responses  the the emulator's keypress server. This way we can automatically respond to all the stuff displayed on screen.

The Travis config has also been updated. Ubuntu Bionic was needed for the emulator to run, so Travis has been updated to use Bionic instead of Xenial. All of the necessary dependencies will be installed.

Lastly, `test_big_tx` on the emulator takes ~15 minutes. This is long enough to cause Travis to timeout. To get around this, we will install the [`travis-wait-enhanced`](https://github.com/crazy-max/travis-wait-enhanced) tool. This will print out an information line every minute so that we don't hit the ["No output was received" timeout](https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received) but still print the test results as the tests run (Travis's recommended `travis_wait` function unfortunately does not print out the test output until the command finishes, making it really hard to know what is going on). Our Travis runs are starting to take quite a long time now.

Closes #277

### CODE REVIEW

**Comment by instagibbs** [comment]
`ledger: test_backup (test_ledger.ledger_test_suite.<locals>.TestLedgerDisabledCommands) ... [Errno 111] Connection refused`

in a bunch of the test timeouts

concept ACK! Would this allow me to still run manual testing on my own device, since f.e., this doesn't emulate Nano X?

File: test/setup_environment.sh [hunk 1]
```diff
@@ -162,6 +162,38 @@ pipenv run make -j$(nproc) kkemu
 find . -name "emulator.img" -exec rm {} \;
 cd ..
 
+# Clone ledger simulator Speculos if it doesn't exist, or update it if it does
+speculos_setup_needed=false
+if [ ! -d "speculos" ]; then
+    git clone --recursive https://github.com/LedgerHQ/speculos.git
```
Comment (instagibbs) [inline test/setup_environment.sh]: we should choose a specific hash to target a commit

Comment (achow101) [reply test/setup_environment.sh]: We don't do that for the other simulators...

**Comment by achow101** [comment]
For some reason the transaction signing tests time out on travis. This doesn't happen when I run it locally. Maybe for now we should skip those tests?

**Comment by instagibbs** [comment]
I'm seeing lots of failures, not simple timeouts?

**Comment by achow101** [comment]
The failures were travis-wait-extended timing out, not travis itself timing out.

**Comment by instagibbs** [comment]
Having issues running this(and actually the other emulators) locally. Investigating. 

concept ACK to merging solution provided they can be run locally reliably.

**Comment by instagibbs** [comment]
tests hanging locally:
`ledger: test_backup (test_ledger.ledger_test_suite.<locals>.TestLedgerDisabledCommands) ...`

investigating along with https://github.com/bitcoin-core/HWI/issues/288

**Comment by instagibbs** [comment]
I was having an import error for `DongleServer`. For some reason re-arranging the definition to before `HIDDongleHIDAPI` causes it to not fail to import, and passes tests.

**Comment by instagibbs** [comment]
it's now working without any modification... I don't understand this. but tACK I guess

**Comment by instagibbs** [comment]
tACK e085f99add575f429763875e1c1af2c28312e393

**Pull request merged**